File: 352/solution.go
package main

import (
	"fmt"
	"math"
	"slices"
	"strings"
)

func main() {
	expect([]int{2, 4, 1, 3, -5, 6}, 72)
	// Extra tests
	expect([]int{3, 1, 8, 0, -3, -5}, 120)
	expect([]int{3, 1, 8, 0, -3, 6}, 144)
}

func expect(arg []int, expected int) {
	result := maxProduct(arg)
	if result != expected {
		var s = strings.ReplaceAll(fmt.Sprint(arg), " ", ", ")
		fmt.Printf("expected maxProduct(%s) == %d, got %d\n", s, expected, result)
	}
}

func maxProduct(list []int) int {
	slices.Sort(list)
	n := len(list)
	a, b, c := list[0], list[1], list[n-1]
	x, y, z := list[n-3], list[n-2], list[n-1]

	return int(math.Max(float64(a*b*c), float64(x*y*z)))
}

File: 352/solution.js
main();

function main() {
  expect([2, 4, 1, 3, -5, 6], 72);
  // Extra tests
  expect([3, 1, 8, 0, -3,-5], 120);
  expect([3, 1, 8, 0, -3, 6], 144);
}

function expect(args, expected) {
  const result = maxProduct(args);
  if (result != expected) {
    console.error(`Expected maxProduct(${args.join(", ")}) == ${expected}, got ${result}`);
  }
}

function maxProduct(list) {
  const sorted = list.toSorted();
  const n = sorted.length;
  // Grab the first two minimum values and the last. The first two might be large
  // negative values.
  const [a, b, c] = [sorted[0], sorted[1], sorted[n-1]];
  // Grab the last two maximum values.
  const [x, y, z] = [sorted[n-3], sorted[n-2], sorted[n-1]];
  return Math.max(a * b * c, x * y * z);
}